Adding accept header to callsites within SignalR

MattyLeslie opened the pull request:
**Adding accept header to callsites within SignalR**


**Summary of the changes**
As mentioned in #47398, many of the `HttpClient` call sites within the SignalR library were missing their appropriate accept headers. This change aims to address that issue. 

**Description**
Changes made include:
1) **Adding** accept header to `NegotiateAsync()` within `HttpConnection` to ensure header is present for negotiation requests.
2) **Adding** accept header to `SendMessage()` within `SendUtils` for the `send` requests.
3) **Adding** accept header to `Poll()` within `LongPollingTransport` for the `get` request.
4) **Test** to ensure first request and subsequent polling requests for `LongPollingTransport` have the correct accept header
5) **Test** to ensure the request for `ServerSentEventsTransport` has the correct accept header
6) **Test** to ensure the negotiate in `HttpConnection` has the correct accept header. 

Fixes #47398 by ensuring the negotiation requests and other HttpClient call sites of SignalR are made with the relevant accept header.

MattyLeslie committed:
Adding accept header within NegotiateAsync() and default headers in CreateHttpClient()

MattyLeslie committed:
Removing defaultRequestHeader for accept and focusing on setting the value on relevant call sites

MattyLeslie committed:
Removing the logic to clear accept headers before allocating the accept header in NegotiateAsync()

MattyLeslie committed:
Adding accept header to LongPolling Get request within the Poll() method

MattyLeslie committed:
Adding the accept header to SendUtils call site.

MattyLeslie committed:
Added the accept header to the LongPolling callsite for the inital polling request.

MattyLeslie committed:
Merge branch 'dotnet:main' into SignalR--Missing-Accept-Header

MattyLeslie committed:
Removing whitespaces

MattyLeslie committed:
Removing comment and fixing whitespace

MattyLeslie committed:
Correcting indentations

MattyLeslie committed:
retrigger checks

MattyLeslie committed:
retrigger checks

MattyLeslie committed:
retrigger checks

MattyLeslie committed:
Merge branch 'dotnet:main' into SignalR--Missing-Accept-Header

MattyLeslie committed:
retrigger checks

MattyLeslie committed:
Merge branch 'dotnet:main' into SignalR--Missing-Accept-Header

BrennanConroy commented:
Are you actually changing anything, or just kicking the build? If it's kicking the build please stop.
Also, could you please look into adding tests, see https://github.com/dotnet/aspnetcore/blob/42bc1a131e745d91a22030aea28af924b424e4e4/src/SignalR/common/Http.Connections/test/ServerSentEventsTests.cs#L34 as an example.

MattyLeslie committed:
Removing default headers

MattyLeslie committed:
adding tests for LongPollingTransport to check accept headers are set accoringly

MattyLeslie committed:
Adding tests for ServerSentEventsTransport to check accept headers are set accoringly

MattyLeslie committed:
Merge branch 'dotnet:main' into SignalR--Missing-Accept-Header

MattyLeslie committed:
Removing the Accept header addition for the SendDeleteRequestMethod()

BrennanConroy commented:
Looks like you're missing negotiate? Otherwise I think that covers all the spots.

MattyLeslie committed:
HttpConnection tests for NegotiateAsync

MattyLeslie committed:
Better method naming

MattyLeslie committed:
Improving LongPollingTransportTests

MattyLeslie committed:
Creating accept header tests for the SendUtils

MattyLeslie committed:
Improving ServerSentEventTests for accept header

BrennanConroy commented:
I'm sorry to see this closed, anything I can help with? It looked really close to finished.

MattyLeslie committed:
Adding an option parameter to HttpConnections contructor used for unit testing. Adding an optional HttpMessageHandler parameter for CreateHttpClient.

MattyLeslie committed:
Addition of Test method to ensure NegotiateAsync() appends the correct header to the Http Request

MattyLeslie commented:
Hello @BrennanConroy,

I've reopened the PR and pushed some new commits. 

**I'm encountering a couple of challenges that I'm currently addressing:**
1) For the `ServerSentEventsTransport.cs` call site, the test hangs if the accept header is missing from the request in the `StartAsync()` method.
2) The test I created for the `NegotiateAsync()` method in `HttpConnection.cs` isn't passing. It seems that not all dependencies are being adequately mocked.

To better test the `NegotiateAsync()` method, I modified the unit testing constructor to accept a nullable `HttpMessageHandler`. This change allows me to mock `HttpMessageHandler` in the `CreateHttpClient()` method, enabling me to verify the outgoing requests made by the `HttpClient` during `NegotiateAsync()`. I hope this change is appropriate.

I welcome any constructive feedback on the testing approach I've implemented for the various call sites. I'm actively working to resolve the issues mentioned above.

BrennanConroy commented:
> To better test the `NegotiateAsync()` method, I modified the unit testing constructor to accept a nullable `HttpMessageHandler`. This change allows me to mock `HttpMessageHandler` in the `CreateHttpClient()` method, enabling me to verify the outgoing requests made by the `HttpClient` during `NegotiateAsync()`. I hope this change is appropriate.

Why not use `HttpConnectionOptions.HttpMessageHandlerFactory`?
https://github.com/dotnet/aspnetcore/blob/0bae45ebe88ad07f69e13a142c93a05b496dc9b4/src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs#L3272

MattyLeslie committed:
Removing changes related to injecting httpMessageHandler

MattyLeslie committed:
Adjusting test to make use of httpmessageHandlerFactory and better practices

MattyLeslie committed:
Updating implementation for NegotiateAsyncAppendsCorrectAcceptHeader

MattyLeslie committed:
Uncommenting the addition of the accept header in NegotiateAsync()

MattyLeslie committed:
Final implementation for checking the correct header is appended to NegotiateAsync

MattyLeslie committed:
Ensuring the TaskCompletionSource is allocated to

MattyLeslie committed:
Merge branch 'dotnet:main' into SignalR--Missing-Accept-Header

MattyLeslie committed:
Merge branch 'SignalR--Missing-Accept-Header' of https://github.com/MattyLeslie/aspnetcore into SignalR--Missing-Accept-Header

MattyLeslie commented:
Thank you very much for the insight @BrennanConroy. I believe I have completed all the necessary tests.

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/HttpConnectionTests.cs:
```diff
@@ -157,4 +162,44 @@ public async Task HttpRequestAndErrorResponseLogged()
+    public async Task NegotiateAsyncAppendsCorrectAcceptHeader()
+    {
+        var testHttpHandler = new TestHttpMessageHandler(false);
+        var negotiateUrlTcs = new TaskCompletionSource<bool>();
```
nit: `TaskCreationOptions.RunContinuationsAsynchronously` for all the `TaskCompletionSource`s

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/LongPollingTransportTests.cs:
```diff
@@ -18,6 +18,7 @@
 using Moq;
 using Moq.Protected;
 using Xunit;
+using System.Net.Http.Headers;
```
Sort headers

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/HttpConnectionTests.cs:
```diff
@@ -17,6 +17,11 @@
 using Xunit;
+using Moq.Protected;
+using System.IO.Pipelines;
+using System.Text;
```
Sort headers

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/LongPollingTransportTests.cs:
```diff
@@ -692,4 +693,45 @@ public async Task SendsDeleteRequestWhenTransportCompleted()
+    [Fact]
+    public async Task StartAsyncSetsAcceptHeaderCorrectly()
+    {
+        var mockHttpHandler = new Mock<HttpMessageHandler>();
```
Why use `Mock` for this test and `TestHttpMessageHandler` for the other one? I'd prefer staying away from Mock in general (even though we use it for some tests).

MattyLeslie replied:
In all honestly, I was just sticking as close to the practices/methodologies found within the same testing classes. I have amended all the tests I wrote to use `TestHttpMessageHandler`. Which I should've originally done.

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/LongPollingTransportTests.cs:
```diff
@@ -692,4 +693,45 @@ public async Task SendsDeleteRequestWhenTransportCompleted()
+            {
+                if (request.Headers.Accept?.Contains(new MediaTypeWithQualityHeaderValue("*/*")) == true)
+                {
+                    responseTaskCompletionSource.SetResult(ResponseUtils.CreateResponse(HttpStatusCode.OK));
```
Isn't this going to be hit twice? Which would throw internally I think?

MattyLeslie replied:
I believe you are correct; the `TaskCompletionSouce` can only be completed once. I have refactored the test with an improved methodology.

BrennanConroy reviewed (commented):
Just some cleanup comments

MattyLeslie committed:
Updating all TaskCompletionSources to run  continuations asynchronously

MattyLeslie committed:
Sorting headers

MattyLeslie committed:
SendUtils test to use testHttpHandler instead of moq

MattyLeslie committed:
ServerSentEvents headers test to use testHttpHandler instead of moq

MattyLeslie committed:
Improvement to PollRequestsContainCorrectAcceptHeader method to effectively handle the intial http request and the requests in Poll()

MattyLeslie committed:
Merge branch 'dotnet:main' into SignalR--Missing-Accept-Header

MattyLeslie commented:
@BrennanConroy, it seems functional tests from MVC are failing due to `httpClient.timeout`. Can I assume this is network related ?

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/HttpConnectionTests.cs:
```diff
@@ -157,4 +162,44 @@ public async Task HttpRequestAndErrorResponseLogged()
+            // ignore connection error
+        }
+
+        Assert.True(negotiateUrlTcs.Task.IsCompleted);
+        var headerWasFound = await negotiateUrlTcs.Task;
```
```suggestion
        var headerWasFound = await negotiateUrlTcs.Task.DefaultTimeout();
```

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/LongPollingTransportTests.cs:
```diff
@@ -692,4 +694,77 @@ public async Task SendsDeleteRequestWhenTransportCompleted()
+                responseTaskCompletionSource.TrySetResult(new HttpResponseMessage(HttpStatusCode.BadRequest));
+            }
+
+            Assert.True(responseTaskCompletionSource.Task.IsCompleted);
+            var response = await responseTaskCompletionSource.Task;
```
```suggestion
            var response = await responseTaskCompletionSource.Task.DefaultTimeout();
```

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/ServerSentEventsTransportTests.cs:
```diff
@@ -409,4 +410,38 @@ public async Task SSETransportThrowsForInvalidTransferFormat(TransferFormat tran
+            await sseTransport.StartAsync(new Uri("http://fakeuri.org"), TransferFormat.Text).DefaultTimeout();
+            await sseTransport.StopAsync().DefaultTimeout();
+
+            Assert.True(responseTaskCompletionSource.Task.IsCompleted);
+            var response = await responseTaskCompletionSource.Task;
```
```suggestion
            var response = await responseTaskCompletionSource.Task.DefaultTimeout();
```

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/SendUtilsTests.cs:
```diff
@@ -0,0 +1,52 @@
+using Moq;
+using Moq.Protected;
+using System.Net.Http.Headers;
+using Microsoft.AspNetCore.InternalTesting;
```
sort

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/LongPollingTransportTests.cs:
```diff
@@ -692,4 +694,77 @@ public async Task SendsDeleteRequestWhenTransportCompleted()
+
+            await firstRequestReceived.Task;
+
+            await Task.Delay(100);
```
We try to avoid adding delays to tests because it creates flakiness. It looks like you can just use `secondRequestReceived` and get rid of `firstRequestReceived`. I don't think canceling the cts does anything?

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/LongPollingTransportTests.cs:
```diff
@@ -692,4 +694,77 @@ public async Task SendsDeleteRequestWhenTransportCompleted()
+
+            await transport.StopAsync();
+
+            if (!responseTaskCompletionSource.Task.IsCompleted)
+            {
+                responseTaskCompletionSource.TrySetResult(new HttpResponseMessage(HttpStatusCode.BadRequest));
+            }
```
Is this needed?

MattyLeslie committed:
Adding default timeouts

MattyLeslie committed:
Sorting using statements and removing Moq

MattyLeslie committed:
Removing delay, only using secondRequestReceived, adding default timeouts and using testUri

MattyLeslie committed:
Merge branch 'SignalR--Missing-Accept-Header' of https://github.com/MattyLeslie/aspnetcore into SignalR--Missing-Accept-Header

MattyLeslie committed:
Adding default time to CompletionSource

BrennanConroy commented:
Thanks @MattyLeslie! Sorry for all the nit-picky comments 😄
